### PR TITLE
external engine: support concurrent read and split into smaller data

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: ref #45719

Problem Summary:

### What is changed and how it works?

- When the external engine contains less than 1000 files, we can open multiple iterators to concurrently read the data in given key ranges.
- When the region size is larger than 2*96MB, generate region jobs per 96MB KV pairs so the large region still have enough concurrency

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  >   

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/backend/external/engine.go [hunk 1]
```diff
@@ -97,9 +101,84 @@ func NewExternalEngine(
...
-func (e *Engine) LoadIngestData(ctx context.Context, start, end []byte) (common.IngestData, error) {
+// MemoryIngestData.DecRef().
+func (e *Engine) LoadIngestData(
+	ctx context.Context,
+	regionRanges []common.Range,
+	outCh chan<- common.DataAndRange,
+) error {
+	// estimate we will open at most 1000 files, so if e.dataFiles is small we can
+	// try to concurrently process ranges.
+	concurrency := int(MergeSortOverlapThreshold) / len(e.dataFiles)
```
[the remark this answers is not in the document]
Comment (lance6716) [reply br/pkg/lightning/backend/external/engine.go:132]: https://github.com/pingcap/tidb/pull/47510/commits/e2cee99630321fc71f4eb1ff9d9525ad14cd0c9e

**Review by Benjamin2037 — APPROVED** [review]
LGTM

File: br/pkg/lightning/backend/external/engine.go [hunk 2]
```diff
@@ -227,13 +324,8 @@ func (e *Engine) SplitRanges(
...
 
-// Close releases the resources of the engine.
-func (e *Engine) Close() error {
-	if e.iter == nil {
-		return nil
-	}
-	return errors.Trace(e.iter.Close())
-}
+// Close implements common.Engine.
+func (e *Engine) Close() error { return nil }
```
Comment (Benjamin2037) [inline br/pkg/lightning/backend/external/engine.go:328]: Doube confirm, external engine close do not need release any resource, right?

Comment (Benjamin2037) [reply br/pkg/lightning/backend/external/engine.go:328]: For the rest lgtm.

Comment (lance6716) [reply br/pkg/lightning/backend/external/engine.go:328]: yes, local engine need to close its pebble DB. And before this PR external engine holds an iterator member that needs to be closed, but in this PR the iterator becomes a local variable that closed in line 144.

**Review by YuJuncen — APPROVED** [review]

**Review by tangenta — APPROVED** [review]

**Pull request merged**